src/UriFactory.php

```diff
@@ -16,4 +36,217 @@ public function createUri(string $uri = ''): UriInterface
+            $https = self::marshalHttpsValue($server['https']);
+        }
+
+        $uri = $uri->withScheme($https ? 'https' : 'http');
+
+        // Set the host
+        [$host, $port] = self::marshalHostAndPort($server, $headers);
+        if (! empty($host)) {
+            $uri = $uri->withHost($host);
+            if (! empty($port)) {
+                $uri = $uri->withPort($port);
+            }
+        }
+
+        // URI path
+        $path = self::marshalRequestPath($server);
+
+        // Strip query string
+        $path = explode('?', $path, 2)[0];
+
+        // URI query
+        $query = '';
+        if (isset($server['QUERY_STRING'])) {
+            $query = ltrim((string) $server['QUERY_STRING'], '?');
+        }
+
+        // URI fragment
+        $fragment = '';
+        if (str_contains($path, '#')) {
+            [$path, $fragment] = explode('#', $path, 2);
+        }
+
+        return $uri
+            ->withPath($path)
+            ->withFragment($fragment)
+            ->withQuery($query);
+    }
+
+    /**
+     * Retrieve a header value from an array of headers using a case-insensitive lookup.
+     *
+     * @template T
+     * @param array<string, string|list<string>> $headers Key/value header pairs
+     * @param T $default Default value to return if header not found
+     * @return string|T
+     */
+    private static function getHeaderFromArray(string $name, array $headers, $default = null)
+    {
+        $header  = strtolower($name);
+        $headers = array_change_key_case($headers, CASE_LOWER);
+        if (! array_key_exists($header, $headers)) {
+            return $default;
+        }
+
+        if (is_string($headers[$header])) {
+            return $headers[$header];
+        }
+
+        return implode(', ', $headers[$header]);
+    }
+
+    /**
+     * Marshal the host and port from the PHP environment.
+     *
+     * @param array<string, string|list<string>> $headers
+     * @return array{string, int|null} Array of two items, host and port,
+     *     in that order (can be passed to a list() operation).
+     */
+    private static function marshalHostAndPort(array $server, array $headers): array
+    {
+        static $defaults = ['', null];
+
+        $host = self::getHeaderFromArray('host', $headers, false);
+        if ($host !== false) {
+            // Ignore obviously malformed host headers:
+            // - Whitespace is invalid within a hostname and break the URI representation within HTTP.
+            //   non-printable characters other than SPACE and TAB are already rejected by HeaderSecurity.
+            // - A comma indicates that multiple host headers have been sent which is not legal
+            //   and might be used in an attack where a load balancer sees a different host header
+            //   than Diactoros.
+            if (! preg_match('/[\\t ,]/', $host)) {
+                return self::marshalHostAndPortFromHeader($host);
```

Comment:
Some of these type check issues require careful checking: in this specific call, for example, `marshalHostAndPortFromHeader` could get better declared types.